### PR TITLE
cmor: bump version and depend on the specific libuuid packages needed by version.

### PR DESCRIPTION
**Pull request by bernhardkaindl**
Fix a problem created by a replacing the `uuid` package with an virtual provider `uuid` which provides different packages with different APIs under one virtual provides.

As `cmor` is not agnostic to the different APIs, depend on the specific packages directly:

cmor versions (`@:3.4.0`) use `uuid_create()` provided by `ossp-uuid`
newer versions (`@3.5.0:`) use `uid_generate()` provided by `libuuid`
Add new versions: Need `json-c` and the `python` prefix directory path

### CODE REVIEW

File: var/spack/repos/builtin/packages/cmor/package.py [hunk 1]
```diff
@@ -13,8 +13,9 @@ class Cmor(AutotoolsPackage):
     standard model experiments."""
 
     homepage = "https://cmor.llnl.gov"
-    url = "https://github.com/PCMDI/cmor/archive/3.4.0.tar.gz"
+    url = "https://github.com/PCMDI/cmor/archive/3.6.1.tar.gz"
 
+    version('3.6.1', sha256='991035a41424f72ea6f0f85653fc13730eb035e63c7dff6ca740aa7a70976fb4')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cmor/package.py:18]: Confirmed the sha256

**Comment by tldahlgren** [comment]
Differing approval/merge to skosukhin 's feedback

**Comment by bernhardkaindl** [comment]
> Differing approval/merge to skosukhin 's feedback

@skosukhin, cmor 3.4.0 uses `uuid_create()` which is not used by any libuuid I could find pernanlly so far. Maybe it's from a very old version of libuuid.

spack's libuuid provides uuid_generate, as do all versions of libuuid that I have, and therefore the build is failing for me:

```yaml
spack-stage-cmor-3.4.0-axmdsi54lryw3tmk47pcm2nwmt7uizcq/spack-src/config.log:
configure:4036: checking for uuid_create in -luuid
/home/kaindlb/spack/spack/.stage/spack-stage-cmor-3.4.0-axmdsi54lryw3tmk47pcm2nwmt7uizcq/spack-src/conftest.c:20:
undefined reference to `uuid_create'
```
cmor-3.6.1 has been updated to use the current uuid_generate:
```yaml
cmor-3.6.1-3f2waesqzceafh4d46t2vrqppoxkpssj/.spack/archived-files/spack-src/config.log:configure:4114:
checking for uuid_generate in -luuid
```
Also spack may want to have the current version of cmor.

**Comment by skosukhin** [comment]
Please, ping me after the 18th of October. I won't be able to take a closer look at this until then.

File: var/spack/repos/builtin/packages/cmor/package.py (hunk 1 above)
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/cmor/package.py:18]: Thanks, added Adam for review since he did most updates of the cmor!

**Review by adamjstewart — COMMENTED** [review]
Only issue I see is that our `uuid` virtual provider is currently not versioned, so `uuid@1.0.3:` doesn't really mean anything. I'm not sure how Spack acts in the case of virtual dependencies without virtual versions. We may need to add versions to:
```console
$ spack providers uuid
uuid:
apple-libuuid  libuuid  ossp-uuid  util-linux-uuid
```

**Comment by bernhardkaindl** [comment]
> Only issue I see is that our `uuid` virtual provider is currently not versioned, so `uuid@1.0.3:` doesn't really mean anything. I'm not sure how Spack acts in the case of virtual dependencies without virtual versions. We may need to add versions to:
> 
> ```
> $ spack providers uuid
> uuid:
> apple-libuuid  libuuid  ossp-uuid  util-linux-uuid
> ```

Doh, that explains everything. `ossp-uuid` provides `uuid_create()` which cmor@:3.4.0 use.
`libuuid` and `util-linux-uuid` provides `uuid_generate()` which is what linux packages get when they link using `-luuid`.

I updated the dependencies accordingly to not depend on `'uuid'` but on `libuuid` (it has no dependencies, `util-linux-uuid` depends on `pkgconfig`). For the older versions, I use a `depends_on('ossp-uuid', when...)`.

With this, 3.4.0 and the newer versions build.

Related, but not to be done in this PR:
@michaelkuhn
```
commit c22141f444861abeaee297a3d92696e9ae94a509
Author: Michael Kuhn <michael.kuhn@ovgu.de>
Date:   Tue Dec 29 17:12:26 2020 +0100

    Introduce virtual provider uuid (#18322)
    
    libuuid is currently contained in util-linux, libuuid and uuid. This
    change introduces a new virtual provider `uuid` and renames the existing
    `uuid` package to `ossp-uuid`.
    
    util-linux's libuuid is provided in the form of a separate package
    util-linux-uuid to make sure that packages depending on uuid and
    util-linux can use a separate uuid implementation, which the concretizer
    does not allow if libuuid is contained in util-linux.
```
It must have been commit which broke the build of `cmor@3.4.0`, as it relies on `uuid_create()`, which `ossp-uuid` provides.

The list of the in total 35 packages to be checked to be affected are:
grep -r "depends_on('uuid'" var/spack/repos/|cut -d/ -f6

I guess a compile test of those should be possible. I've done other much much larger compile test runs as part of checking if the autoconf commits we've just merged break nothing.

I'll consult that with Michael outside of this PR.

**Comment by michaelkuhn** [comment]
Sorry for breaking cmor. I remember rebuilding packages for the uuid change but it might be that cmor slipped through the cracks somehow. Just one comment: I would suggest to use `linux-util-uuid` instead of `libuuid` because the latter is severely outdated. All Linux distributions that I've checked use the util-linux version nowadays.

**Review by alalazo — APPROVED** [review]

**Pull request merged**